### DISCUSSION TITLE
Any info on DU timeline?

### DISCUSSION THREAD

**Discussion by grandaspanna** [SpenceKonde/DxCore#414] [General]
Would love to have native USB - has anyone seen any hints as to when they might appear?

**Comment by fschneidereit** [comment]
> Would love to have native USB - has anyone seen any hints as to when they might appear?

I'd like to know, too. The only hint I've seen recently (October 2022) is that pymcuprog has now support for avr64du28 and avr64du32 devices (see [https://github.com/microchip-pic-avr-tools/pymcuprog/blob/main/pymcuprog/deviceinfo/devices/avr64du28.py](url), for example). This kind of confirms they are still in the pipeline, and these could be the initial configurations released.

**Comment by SpenceKonde** [comment]
Well well, that is interesting indeed! I had begun to suspect that they'd decided not to DU the DU. That things referencing it are appearing seems a strong indicator that that's not the case. Given that they've been averaging 1 product family per year since 2020 (DA/DB in 2020, tiny2 in 2021, DD last year and that the EA is shipping next week (well, kinda), I would not expect it before the coming of the new year - Q1-Q2 of 2024 would be where I would be placing my bets - but I can't say I'm much of a wagering man.

EA is shipping  next week, though quantity may be limited before the fall, and with punishing constraints: you've gotta buy a  whole tray of 250 or 490, or take the dregs: AVR64EA28s in SSOP.  There is only one family of parts so far where any 28-pin parts are a winner, and that's the AVR64DD28 - because stepping up from 20 to 28 pins gives them back the low half of PORTD, which wouldn't matter, except that because the 20-pin parts didn't have the low half of PORTD, so the VDDIO2 pin took PC0 instead of PD0 away, reducing the number of MVIO pins to 3 (the net gain in pins from 20 to 28 is + 3 PORTD pins (which aren't very talented pins), + 1 PORTC pin (with MVIO and a part of several serial/SPI mappings), and +2 PORTF pins (which are undoubtedly the gem of the DD20->28 upgrade, because they can output PWM, and let you have a 32k crystal without burning PA0 and PA1 (which are very useful pins, though the most obvious impossibility precluded by a lack of PF0/PF1, an inability to use an RTC crystal and HF crystal simultaneously, is of limited usefulness) + 1 pair power pins). The only reason this stands out like it does on the double D's is that unlike earlier part families, there's a QFN of the 28-pin part - and it's sold at a price that's hard to argue with. But you know what there isn't? An AVR64DD20 in QFN. 

I filed the plastic top off of an AVR32DD20 and found that the die looked to be nearly as big as the package, implying that they don't made AVR64DD20's in the 3mm x 3mm QFN20 because they can't make the die small enough to fit. I don't recommend doing this as a way of accessing the die if you don't have some files you don't care about (I do - got a giant bag of files for either $1 or free years ago). Now these files obviously weren't new, but still - considering how little material you have to remove to expose the die on a QFN20, I'd expect to get more than through more than 1 chip before declaring a file to be "too dull to use" and trashing it. It appears that there's something abrasive in that plastic that IC packages are made from...

**Reply by grandaspanna** [reply]
A bit OT, but I saw a post somewhere today on using IR inspection to look inside some chip packages. They wanted a non-destructive way to explore chip implementation.

Found it: https://www.hackster.io/news/andrew-bunnie-huang-s-iris-lets-you-peer-at-chips-inner-workings-with-a-cheap-digital-camera-07362fb653dd

**Reply by SpenceKonde** [reply]
Sadly they need the die exposed from the back for that to work as they mention in passing at the bottom, when up until that point it appeared like it gave you x-ray vision without the ionizing radiation (x-ray vision sounds great, but it wouldn't actually be much fun. but you'd eventually get tired of electronics failing whenever you came near, and everyone around you dying of radiation sickness (you'd have to stay moving - and not look at anyone wrong.

**Comment by fschneidereit** [comment]
@SpenceKonde

Thanks for the insight, and your work on DxCore, of course. I was suspecting they dropped the DU, too. Talking about limited quantities, I'm really thankful they still do PDIP variants because these are the ones that were always in stock. The component shortage went so far, I had a small non-commerical design in 2022 around a AVR32DB28 that I did completely in THT. If space is not a constraint and parts are available, why not? Now for another project, I'm eager to get my hands on DU parts because it would make an external USB/UART-Bridge obsolete.

**Comment by SpenceKonde** [comment]
mmm, it doesn't entirely obsolete external USB-serial adapters. They always have one advantage: There's pretty much nothing you can do in software that will make a CH340 or FT232 difficult to recover. It's always easy to do something that makes an MCU with native USB disappear from USB within milliseconds of starting to run the application... There will be a wait before anyone has written a USB bootloader anyway. According to that old product brief, the DUs aren't going to be super exciting parts other than the native USB... 

IMHO the thing that's exciting about native USB is that you can make USB-capable devices that don't present as serial ports :-P I still see people using the classic tinies like t88/t85, because VUSB (with many many caveats) is well exercised there (two level interrupts and performance improvements mean that modern AVRs would likely be able to do bitbanged USB without gunning all interrupts). Though the classic definitely could could never hope to have USB functionality with interrupts enabled it wasn't isn't looks to me like a very aggressively written USB interrupt given the priority slot would be able to do bitbanged USB. If not at 24 MHz, then with a modest overclock (these parts will sometimes run at 48 MHz, which is decidedly immodest). It wouldn't be easy, nor would it be as good as the DU will be,  but I think it could probably be done.

**Comment by SpenceKonde** [comment]
Oh, and if you haven't caught on, DU-series is confirmed by my guy on the inside, he said "hopefully by the end of the year" and recognized the importance to the AVR product line of getting it out ASAP - though I later realized he didn't specify which year....

What is more surprising is that EB headers have beaten the DU headers into an ATpack!

**Comment by XFer012** [comment]
Fingers crossed on the DU; that would put an end for me having to deal with PICs for good

**Comment by FerGT50** [comment]
Any news? Xmas is coming, a DU would be a nice present from Microchip :)

**Comment by SpenceKonde** [comment]
Hey, they said by the "end of the year" - they didn't specify which one!

The EB has managed an astonishing rate of progress! If it's not available
now, it will be soon - so it has made the end of year. It must be that
new PLL.. Not as exciting of a part as the DU though.It was confirmed that
there are some delays from the development of the USB stack, but it is
continuing to progress, and it sounds like there's more of a sense of
optimism.

**Comment by leo60228** [comment]
The latest AVR-Dx pack added DU parts. According to a Microchip employee on the avrdude repo, parts will be available "not before February."

**Comment by grandaspanna** [comment]
Availability next month for some parts. https://www.microchip.com/en-us/products/microcontrollers-and-microprocessors/8-bit-mcus/avr-mcus/avr-du

**Comment by grandaspanna** [comment]
But darn it, a 48-pin variant doesn't look close :-(

**Comment by sstaub** [comment]
When can we expect support by dxcore?

**Comment by tangiblewaves** [comment]
+1 for DU support; I'd love to see DX Core supporting them. Just ordered a few AVR64DU32 :-)))

**Comment by SpenceKonde** [comment]
DU support is near - except without any support for the native USB stuff; I'm hoping someone will be willing to help out with that, because we'll never get USB support if it's just me.

**Comment by mabartibin** [comment]
Oh dear.  I came here to ask whether the core supports USB on the DU parts (I was thinking of MIDI…) but it seems I shouldn’t count on those parts quite yet.  And I’m afraid I lack both the knowhow _and_ the spare time to help with this.

**Comment by leo60228** [comment]
There's an open LUFA PR for these parts, which is probably a workable base for proper USB support: https://github.com/abcminiuser/lufa/pull/187
